File: python_roguelike/core/ai/i_player_agent.py
from abc import ABC, abstractmethod
from dataclasses import dataclass
from ...data.enums import CombatActionType, ShopActionType


@dataclass
class CombatDecision:
    type: CombatActionType = CombatActionType.EndTurn
    hand_index: int = -1
    target_index: int = -1

    @staticmethod
    def end_turn() -> "CombatDecision":
        return CombatDecision(type=CombatActionType.EndTurn)

    @staticmethod
    def play(hand_index: int, target_index: int) -> "CombatDecision":
        return CombatDecision(type=CombatActionType.PlayCard, hand_index=hand_index, target_index=target_index)


@dataclass
class ShopDecision:
    type: ShopActionType = ShopActionType.Leave
    shop_index: int = -1

    @staticmethod
    def leave() -> "ShopDecision":
        return ShopDecision(type=ShopActionType.Leave)

    @staticmethod
    def buy_card(index: int) -> "ShopDecision":
        return ShopDecision(type=ShopActionType.BuyCard, shop_index=index)

    @staticmethod
    def buy_relic(index: int) -> "ShopDecision":
        return ShopDecision(type=ShopActionType.BuyRelic, shop_index=index)


class IPlayerAgent(ABC):
    @abstractmethod
    def choose_map_node(self, run) -> int:
        pass

    @abstractmethod
    def get_combat_decision(self, run) -> CombatDecision:
        pass

    @abstractmethod
    def choose_event_option(self, run) -> int:
        pass

    @abstractmethod
    def get_shop_decision(self, run) -> ShopDecision:
        pass

    @abstractmethod
    def choose_card_reward(self, run) -> int:
        pass

File: src/Roguelike.Core/Core/AI/IPlayerAgent.cs
using Roguelike.Core;

namespace Roguelike.Core.AI
{


    public enum CombatActionType
    {
        PlayCard,
        EndTurn
    }


    public struct CombatDecision
    {
        public CombatActionType Type;
        public int HandIndex;
        public int TargetIndex;

        public static CombatDecision EndTurn() => new CombatDecision { Type = CombatActionType.EndTurn };
        public static CombatDecision Play(int handIndex, int targetIndex) =>
            new CombatDecision { Type = CombatActionType.PlayCard, HandIndex = handIndex, TargetIndex = targetIndex };
    }


    public enum ShopActionType
    {
        BuyCard,
        BuyRelic,
        Leave
    }


    public struct ShopDecision
    {
        public ShopActionType Type;
        public int ShopIndex;

        public static ShopDecision Leave() => new ShopDecision { Type = ShopActionType.Leave };
        public static ShopDecision BuyCard(int index) => new ShopDecision { Type = ShopActionType.BuyCard, ShopIndex = index };
        public static ShopDecision BuyRelic(int index) => new ShopDecision { Type = ShopActionType.BuyRelic, ShopIndex = index };
    }


    public interface IPlayerAgent
    {


        int ChooseMapNode(GameRun run);


        CombatDecision GetCombatDecision(GameRun run);


        int ChooseEventOption(GameRun run);


        ShopDecision GetShopDecision(GameRun run);


        int ChooseCardReward(GameRun run);
    }
}
